# 1장 파이썬 머신러닝 이해  



[머신러닝 완벽 가이드] 책을 기반으로 제가 정리하고 요약한 내용입니다.  
1장은 머신러닝의 기초인 넘파이와 판다스에 대한 개념입니다.

## 넘파이
넘파이는 배열 기반의 연산은 물론이고 다양한 데이터 핸들링 기능을 제공합니다.

In [1]:
import numpy as np # 넘파이 모듈 임포트

### ndarray 데이터 타입

In [2]:
array1 = np.array([1,2,3]) 
# 넘파이의 array()함수는 파이썬의 리스트와 같은 인자를 ndarray로 변환하는 기능 수행
print('array1 type:',type(array1))
print('array1 type 형태:',array1.shape)

array2 = np.array([[1,2,3],
                   [4,5,6]])
print('array2 type:',type(array2))
print('array2 type 형태:',array2.shape)

array1 type: <class 'numpy.ndarray'>
array1 type 형태: (3,)
array2 type: <class 'numpy.ndarray'>
array2 type 형태: (2, 3)


In [3]:
list1 = [1,2,3]
print(type(list1))
array1 = np.array(list1)
print(type(array1))
print(array1.dtype) #ndarray내의 데이터타입은 .dtype 으로 확인 가능

<class 'list'>
<class 'numpy.ndarray'>
int32


In [4]:
list2 = [1,2,'test']
array2 = np.array(list2)
print(array2.dtype)

list3 = [1,2,3.1]
array3 = np.array(list3)
print(array3.dtype) 

<U11
float64


다른 데이터 유형이 ndarray로 변경되면 데이터 크기가 더 큰 데이터 타입으로 형 변환이 적용됩니다.

In [5]:
list3 = [1, 2, 3]
array4 =np.array(list3)
print(array4.dtype)
array4 =array4.astype(float)
print(array4.dtype)

int32
float64


ndarray 내 데이터값의 타입 변경은 astype()을 이용해 가능합니다.

### ndarray 생성

In [6]:
arange = np.arange(10)
print(arange)
print(arange.dtype)
print(arange.shape)

[0 1 2 3 4 5 6 7 8 9]
int32
(10,)


arange()는 0부터 함수 인자 값 -1까지의 값을 순차적으로 ndarray의 데이터값으로 변환해줍니다.  
start, stop을 지정해서 원하는 값부터 시작한 연속 값을 부여할 수 있습니다.

In [7]:
zeros = np.zeros((3,2))
print(zeros)
print(zeros.dtype)
print(zeros.shape)

[[0. 0.]
 [0. 0.]
 [0. 0.]]
float64
(3, 2)


zeros()는 함수 인자로 튜플 형태의 shape값을 입력하면 모든 값을 0으로 채운 해당 shape를 가진 ndarray를 반환합니다. 이와 유사하게 ones()는 모든 값을 1로 채운 ndarray를 반환합니다. default는 float입니다.

In [8]:
arange = np.arange(10)
print(arange)
print(arange.reshape(2,5))
print(arange.reshape(5,-1))

[0 1 2 3 4 5 6 7 8 9]
[[0 1 2 3 4]
 [5 6 7 8 9]]
[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]


reshape()는 원하는 형태의 ndarray로 변환해주며, 지정된 사이즈로 변경이 불가능하면 오류가 발생합니다.  
-1를 인자로 사용하면 원래 ndarray와 호환되는 새로운 shape로 변환해줍니다.

### ndarray 데이터 세트 선택 - 인덱싱

In [9]:
array1 = np.arange(start=1,stop=10)
array1

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [10]:
array1[2]
print(array1[2],array1[2].dtype)
array1[:5]
print(array1[array1[:5]])
array1[:-1]
print(array1[:-1])

3 int32
[2 3 4 5 6]
[1 2 3 4 5 6 7 8]


인덱스는 0부터 시작하므로 [2]는 3번째 인덱스 위치의 데이터 값을 의미합니다.  
타입은 ndarray타입이 아니고 ndarray 내의 데이터값을 의미합니다.  
-1은 맨 뒤 데이터값을 의미하며 -2는 맨 뒤에서 두 번째에 있는 데이터값을 말합니다.

In [11]:
array1d = np.arange(start=1,stop=10)
array2d = array1d.reshape(3,3)
print(array2d)
print(array2d[0,0])
print(array2d[1,1])
print(array2d[2,2])

[[1 2 3]
 [4 5 6]
 [7 8 9]]
1
5
9


다차원 ndarray에서 단일 값 추출 경우에는 콤마(,)로 분리된 로우와 칼럼 위치의 인덱스를 통해 접근합니다.

In [12]:
array = np.arange(start=1,stop=10)
print(array)
print(array[:3])
print(array[3:])
print(array[:-1])

[1 2 3 4 5 6 7 8 9]
[1 2 3]
[4 5 6 7 8 9]
[1 2 3 4 5 6 7 8]


':' 기호 앞에 시작 인덱스를 생략하면 자동으로 맨 처음 인덱스인 0으로 간주
':' 기호 뒤에 종료 인덱스를 생략하면 자동으로 맨 마지막 인덱스로 간주
':' 기호 앞/뒤에 시작/종료 인덱스를 생략하면 자동으로 맨처음/맨 마지막 인덱스로 간주

In [13]:
array = np.arange(start=1,stop=10)
array = array.reshape(3,3)
print(array)
print(array[0:2,0:2])
# print(array[1:2,1:3])
print(array[1:3,:2])

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1 2]
 [4 5]]
[[4 5]
 [7 8]]


2차원 ndarray에서도 1차원 ndarray에서의 슬라이싱과 유사하며, 콤마(,)로 로우와 칼럼 인덱스를 지칭

In [14]:
array = np.arange(1,10)
array = array.reshape(3,3)
print(array)
print(array[[0,1],2])
print(array[[0,1],:2])


[[1 2 3]
 [4 5 6]
 [7 8 9]]
[3 6]
[[1 2]
 [4 5]]


리스트나 ndarray로 인덱스 집합을 지정하면 해당 위치의 인덱스에 해당하는 ndarray를 반환합니다.

In [15]:
array = np.arange(1,10)
print(array)
print(array>5)
print(array[array>5])

[1 2 3 4 5 6 7 8 9]
[False False False False False  True  True  True  True]
[6 7 8 9]


조건 필터링과 검색을 동시에 할 수 있는 방식을 불린 인덱싱 방식이라고 합니다.

### 행렬의 정렬

In [16]:
array = np.array([3,1,9,5])
print(array)
print(np.sort(array))
print(array)
array1 = array.sort()
print(array1)
print(array)

array = np.array([3,1,9,5])
print("내림차순:",np.sort(array)[::-1])

[3 1 9 5]
[1 3 5 9]
[3 1 9 5]
None
[1 3 5 9]
내림차순: [9 5 3 1]


np.sort()는 원본 행렬에는 영향을 주지않고 정렬된 행렬만을 반환하며,  
ndarray.sort()는 원 행렬 자체를 정렬한 형태로 변환하며, 반환값은 None입니다.
모두 오름차순으로 원소를 정렬하며 내림차순으로 정렬하기 위해서는 [::-1]을 적용합니다.

In [17]:
array = np.array([[3,5],
                  [5,4]])
print(array)
print("axis=0:\n",np.sort(array,axis=0))
print("axis=1:\n",np.sort(array,axis=1))

[[3 5]
 [5 4]]
axis=0:
 [[3 4]
 [5 5]]
axis=1:
 [[3 5]
 [4 5]]


2차원 이상 행렬인 경우 axis 축 값 설정을 통해 로우 방향, 또는 칼럼 방향으로 정렬을 수행할 수 있습니다.

In [18]:
array = np.array([3,1,9,5])
print(array)
print("argsort:\n",np.argsort(array))

[3 1 9 5]
argsort:
 [1 0 3 2]


원본 행렬의 원소에 대한 인덱스를 필요로 할 때 np.argsort를 이용합니다.

### 선형대수 연산 - 행렬 내적과 전치 행렬 구하기

In [19]:
A= np.array([[1,2,3],[4,5,6]])
B= np.array([[7,8],[9,10],[11,12]])

np.dot(A,B)


array([[ 58,  64],
       [139, 154]])

행렬 내적은 행렬 곱이며, 두 행렬의 내적은 np.dot()을 이용해 계산합니다.  
주의해야할 특성으로는 왼쪽 행렬의 열 개수와 오른쪽 행렬의 행 개수가 동일해야 합니다.

In [20]:
print(A)
print(A.T)

[[1 2 3]
 [4 5 6]]
[[1 4]
 [2 5]
 [3 6]]


행렬에서 행과 열의 위치를 교환한 원소로 구성한 행렬을 전치행렬이라하며, transpose 또는 T를 이용해 구할 수 있습니다.

## 판다스

판다스의 핵심 객체는 데이터프레임으로, 여러 개의 행과 열로 이뤄진 2차원 데이터를 담는 데이터 구조체입니다. 데이터프레임을 이해하기 전에 다른 중요 객체인 인덱스와 시리즈를 이해하는 것이 중요하며 인덱스는 개별 데이터를 식별하는 '키'값입니다. 시리는 칼럼이 하나뿐인 데이터 구조체이며, 데이터프레임은 칼럼이 여러 개인 데이터 구조체입니다.

In [21]:
import pandas as pd # 판다스 모듈 임포트

In [25]:
df = pd.read_csv('train.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

데이터프레임으로 로딩할 수 있는 API로  
read_csv( ), read_table( ), read_fwf( )가 존재하며, csv()와 table()의 차이는 필드 구분 문자가 콤마(,)냐, 탭("\t")이냐의 차이입니다.  
가령 탭으로 필드가 구분돼 있다면 read_csv('파일명', sep='\t')처럼 사용 가능하며, sep를 생략하면 자동으로 콤마로 할당합니다.

read_csv('파일명')에서 '파일명'은 파이썬 실행 파일이 있는 디렉터리와 동일한 디렉터리에 존재해야하며, 같은 디렉터리에 존재하지않는 경우에는 해당 파일에 있는 경로를 입력해주면 불러올 수 있습니다.


In [29]:
df.head(3)

(891, 12)

데이터프레임의 모든 데이터를 한번에 표출하면 메모리가 많이 사용되므로 head()를 이용해 n개의 로우를 반환합니다.default는 5개이며 head(3)은 맨 앞 3개의 로우를 반환합니다.

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


info()를 통해 총 데이터 건수 891개, 컬럼수는 12개인 것을 확인 할 수 있습니다.  
컬럼마다 데이터 형태를 확인할 수 있으며, 어떤 컬럼의 몇개의 널값이 존재하는지 확인할 수 있습니다.  
메모리 사용량도 확인 가능합니다.

In [35]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

- count: 널값이 아닌 데이터 건수  
- mean : 전체 데이터 평균값  
- std : 표준편차  
- min,max : 최소,최대값  
- 25%, 50%, 75% : percentile 값

In [43]:
df['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

칼럼의 값이 어떠한 분포로 구성되어 있는지 확인하기 위해서는 value_counts()를 사용합니다. 지정된 칼럼의 데이터값 건수를 반환하며, 분포도를 확인하기 유용합니다.